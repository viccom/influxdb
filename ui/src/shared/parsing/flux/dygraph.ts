// Libraries
import {get} from 'lodash'

// Utils
import {
  spreadTables,
  SeriesDescription,
} from 'src/shared/parsing/flux/spreadTables'

// Types
import {FluxTable} from 'src/types'
import {DygraphValue} from 'src/external/dygraph'

export interface FluxTablesToDygraphResult {
  labels: string[]
  dygraphsData: DygraphValue[][]
  seriesDescriptions: SeriesDescription[]
}

export const fluxTablesToDygraph = (
  tables: FluxTable[]
): FluxTablesToDygraphResult => {
  const {table, seriesDescriptions} = spreadTables(tables)
  const labels = seriesDescriptions.map(d => d.key)

  labels.sort()

  const dygraphsData = Object.keys(table).map(time => [
    new Date(time),
    ...labels.map(label => get(table, [time, label], null)),
  ])

  dygraphsData.sort((a, b) => (a[0] as any) - (b[0] as any))

  return {dygraphsData, labels: ['time', ...labels], seriesDescriptions}
}
