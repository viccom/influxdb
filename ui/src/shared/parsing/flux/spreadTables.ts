import {FluxTable} from 'src/types'

export interface SeriesDescription {
  // A key identifying a unique (column, table, result) triple for a particular
  // Flux response—i.e. a single time series
  key: string
  // The name of the column that this series is extracted from (typically this
  // is `_value`, but could be any column)
  valueColumnName: string
  // The index of the column that this series was extracted from
  valueColumnIndex: number
  // The names and values of columns in the group key, plus the result name.
  // This provides the data for a user-recognizable label of the time series
  metaColumns: {
    [columnName: string]: string
  }
}

interface SpreadTablesResult {
  seriesDescriptions: SeriesDescription[]
  table: {
    [time: string]: {[seriesKey: string]: number}
  }
}

// Given a collection of `FluxTable`s parsed from a single Flux response,
// `spreadTables` will place each unique series found within the response into
// a single table, indexed by time. This data munging operation is often
// referred to as as a “spread”, “cast”, “pivot”, or “unfold”.
export const spreadTables = (tables: FluxTable[]): SpreadTablesResult => {
  const result: SpreadTablesResult = {
    table: {},
    seriesDescriptions: [],
  }

  for (const table of tables) {
    const header = table.data[0]

    if (!header) {
      continue
    }

    const seriesDescriptions = getSeriesDescriptions(table)
    const timeIndex = getTimeIndex(header)

    for (let i = 1; i < table.data.length; i++) {
      const row = table.data[i]
      const time = row[timeIndex]

      for (const {key, valueColumnIndex} of seriesDescriptions) {
        if (!result.table[time]) {
          result.table[time] = {}
        }

        result.table[time][key] = Number(row[valueColumnIndex])
      }
    }

    result.seriesDescriptions.push(...seriesDescriptions)
  }

  return result
}

const EXCLUDED_SERIES_COLUMNS = new Set([
  '_time',
  'result',
  'table',
  '_start',
  '_stop',
  '',
])

const NUMERIC_DATATYPES = new Set(['double', 'long', 'int', 'float'])

const getSeriesDescriptions = (table: FluxTable): SeriesDescription[] => {
  const seriesDescriptions = []
  const header = table.data[0]

  for (let i = 0; i < header.length; i++) {
    const columnName = header[i]
    const dataType = table.dataTypes[columnName]

    if (EXCLUDED_SERIES_COLUMNS.has(columnName)) {
      continue
    }

    if (table.groupKey[columnName]) {
      continue
    }

    if (!NUMERIC_DATATYPES.has(dataType)) {
      continue
    }

    const key = Object.entries(table.groupKey).reduce(
      (acc, [k, v]) => acc + `[${k}=${v}]`,
      `${columnName}[result=${table.result}]`
    )

    seriesDescriptions.push({
      key,
      valueColumnName: columnName,
      valueColumnIndex: i,
      metaColumns: {
        ...table.groupKey,
        result: table.result,
      },
    })
  }

  return seriesDescriptions
}

const getTimeIndex = header => {
  let timeIndex = header.indexOf('_time')

  if (timeIndex >= 0) {
    return timeIndex
  }

  timeIndex = header.indexOf('_start')
  if (timeIndex >= 0) {
    return timeIndex
  }

  timeIndex = header.indexOf('_end')
  if (timeIndex >= 0) {
    return timeIndex
  }

  return -1
}
