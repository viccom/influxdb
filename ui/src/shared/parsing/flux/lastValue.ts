import {spreadTables} from 'src/shared/parsing/flux/spreadTables'

import {FluxTable} from 'src/types'

export const lastValue = (tables: FluxTable[]): number => {
  if (tables.every(table => !table.data.length)) {
    return null
  }

  const {table, seriesDescriptions} = spreadTables(tables)
  const seriesKeys = seriesDescriptions.map(d => d.key)
  const times = Object.keys(table)

  times.sort()
  seriesKeys.sort()

  const lastTime = times[times.length - 1]
  const firstSeriesKey = seriesKeys[0]

  return table[lastTime][firstSeriesKey]
}
