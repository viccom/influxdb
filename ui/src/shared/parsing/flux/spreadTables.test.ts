import {parseResponse} from 'src/shared/parsing/flux/response'
import {spreadTables} from 'src/shared/parsing/flux/spreadTables'

describe('spreadTables', () => {
  test('it spreads multiple series into a single table', () => {
    const resp = `#group,false,false,false,false,false,false,true,true,true
#datatype,string,long,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339,long,string,string,string
#default,max,,,,,,,,
,result,table,_start,_stop,_time,_value,_field,_measurement,host
,,0,2018-12-10T18:21:52.748859Z,2018-12-10T18:30:00Z,2018-12-10T18:29:58Z,4906213376,active,mem,oox4k.local
,,0,2018-12-10T18:30:00Z,2018-12-10T19:00:00Z,2018-12-10T18:54:08Z,5860683776,active,mem,oox4k.local
,,0,2018-12-10T19:00:00Z,2018-12-10T19:21:52.748859Z,2018-12-10T19:11:58Z,5115428864,active,mem,oox4k.local

#group,false,false,false,false,false,false,true,true,true
#datatype,string,long,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339,long,string,string,string
#default,min,,,,,,,,
,result,table,_start,_stop,_time,_value,_field,_measurement,host
,,0,2018-12-10T18:21:52.748859Z,2018-12-10T18:30:00Z,2018-12-10T18:29:48Z,4589981696,active,mem,oox4k.local
,,0,2018-12-10T18:30:00Z,2018-12-10T19:00:00Z,2018-12-10T18:40:18Z,4318040064,active,mem,oox4k.local
,,0,2018-12-10T19:00:00Z,2018-12-10T19:21:52.748859Z,2018-12-10T19:11:58Z,4131692544,active,mem,oox4k.local

`

    const result = spreadTables(parseResponse(resp))

    expect(result).toMatchSnapshot()
  })
})
