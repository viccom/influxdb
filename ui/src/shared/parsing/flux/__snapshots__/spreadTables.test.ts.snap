// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spreadTables it spreads multiple series into a single table 1`] = `
Object {
  "seriesDescriptions": Array [
    Object {
      "key": "_value[result=max][_field=active][_measurement=mem][host=oox4k.local]",
      "metaColumns": Object {
        "_field": "active",
        "_measurement": "mem",
        "host": "oox4k.local",
        "result": "max",
      },
      "valueColumnIndex": 6,
      "valueColumnName": "_value",
    },
    Object {
      "key": "_value[result=min][_field=active][_measurement=mem][host=oox4k.local]",
      "metaColumns": Object {
        "_field": "active",
        "_measurement": "mem",
        "host": "oox4k.local",
        "result": "min",
      },
      "valueColumnIndex": 6,
      "valueColumnName": "_value",
    },
  ],
  "table": Object {
    "2018-12-10T18:29:48Z": Object {
      "_value[result=min][_field=active][_measurement=mem][host=oox4k.local]": 4589981696,
    },
    "2018-12-10T18:29:58Z": Object {
      "_value[result=max][_field=active][_measurement=mem][host=oox4k.local]": 4906213376,
    },
    "2018-12-10T18:40:18Z": Object {
      "_value[result=min][_field=active][_measurement=mem][host=oox4k.local]": 4318040064,
    },
    "2018-12-10T18:54:08Z": Object {
      "_value[result=max][_field=active][_measurement=mem][host=oox4k.local]": 5860683776,
    },
    "2018-12-10T19:11:58Z": Object {
      "_value[result=max][_field=active][_measurement=mem][host=oox4k.local]": 5115428864,
      "_value[result=min][_field=active][_measurement=mem][host=oox4k.local]": 4131692544,
    },
  },
}
`;
